### PR TITLE
W-12689965

### PR DESCRIPTION
**Pull request by damianpedra**
Add GooglePubSub Bindings:
Add Model and Interfaces
Add parsing
Add rendering
Add testing

### CODE REVIEW

File: amf-api-contract/shared/src/main/scala/amf/apicontract/client/platform/model/domain/bindings/googlepubsub/GooglePubSubChannelBinding.scala [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  override def bindingVersion: StrField = _internal.bindingVersion
+
+  def labels: ObjectNode = _internal.labels
+  def messageRetentionDuration: StrField = _internal.messageRetentionDuration
+  def messageStoragePolicy: GooglePubSubMessageStoragePolicy = _internal.messageStoragePolicy
+  def schemaSettings: GooglePubSubSchemaSettings = _internal.schemaSettings
+  def topic: StrField = _internal.topic
+
+
+  def withLabels(labels: NodeShape): this.type = {
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/client/platform/model/domain/bindings/googlepubsub/GooglePubSubChannelBinding.scala]: Should be ObjectNode, no NodeShape

File: amf-api-contract/shared/src/main/scala/amf/apicontract/client/platform/model/domain/bindings/googlepubsub/GooglePubSubMessageBinding.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+case class GooglePubSubMessageBinding(override private[amf] val _internal: InternalGooglePubSubMessageBinding)
+  extends MessageBinding
+  with BindingVersion {
+  @JSExportTopLevel("GooglePubSubMessageBinding")
+  def this() = this(InternalGooglePubSubMessageBinding())
+
+  override protected def bindingVersion: StrField = _internal.bindingVersion
+  def attributes: ObjectNode = _internal.attributes
+  def orderingKey: StrField = _internal.orderingKey
+  def schema: InternalGooglePubSubSchemaDefinition = new GooglePubSubSchemaDefinition(_internal.schema)
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/client/platform/model/domain/bindings/googlepubsub/GooglePubSubMessageBinding.scala]: Should return GooglePubSubSchemaDefinition, not InternalGooglePubSubSchemaDefinition

File: amf-api-contract/shared/src/main/scala/amf/apicontract/client/platform/model/domain/bindings/googlepubsub/GooglePubSubMessageBinding.scala [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+  def withAttributes(attributes: ObjectNode): this.type = {
+    _internal.withAttributes(attributes)
+    this
+  }
+  def withOrderingKey(orderingKey: String): this.type = {
+    _internal.withOrderingKey(orderingKey)
+    this
+  }
+  def withSchema(schema: InternalGooglePubSubSchemaDefinition): this.type = {
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/client/platform/model/domain/bindings/googlepubsub/GooglePubSubMessageBinding.scala]: Should receive GooglePubSubSchemaDefinition, not InternalGooglePubSubSchemaDefinition

File: amf-api-contract/shared/src/main/scala/amf/apicontract/client/scala/model/domain/bindings/googlepubsub/GooglePubSubChannelBinding.scala [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    with BindingVersion
+    with Key {
+  override protected def bindingVersionField: Field          = BindingVersion
+  def labels: ObjectNode                                      = fields.field(Labels)
+  def messageRetentionDuration: StrField                     = fields.field(MessageRetentionDuration)
+  def messageStoragePolicy: GooglePubSubMessageStoragePolicy = fields.field(MessageStoragePolicy)
+  def schemaSettings: GooglePubSubSchemaSettings             = fields.field(SchemaSettings)
+  def topic: StrField                                        = fields.field(Topic)
+
+  def withLabels(labels: NodeShape): this.type = set(Labels, labels)
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/client/scala/model/domain/bindings/googlepubsub/GooglePubSubChannelBinding.scala]: Should be ObjectNode, no NodeShape

File: amf-api-contract/shared/src/main/scala/amf/apicontract/client/scala/model/domain/bindings/googlepubsub/GooglePubSubChannelBinding.scala [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+import amf.core.client.scala.model.StrField
+import amf.core.client.scala.model.domain.{DomainElement, Linkable, ObjectNode}
+import amf.core.internal.metamodel.Field
+import amf.core.internal.parser.domain.{Annotations, Fields}
+import amf.shapes.client.scala.model.domain.{Key, NodeShape}
+class GooglePubSubChannelBinding(override val fields: Fields, override val annotations: Annotations)
+    extends ChannelBinding
+    with BindingVersion
+    with Key {
+  override protected def bindingVersionField: Field          = BindingVersion
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/client/scala/model/domain/bindings/googlepubsub/GooglePubSubChannelBinding.scala:17]: Should return StrField, not Field

File: amf-api-contract/shared/src/main/scala/amf/apicontract/client/scala/model/domain/bindings/googlepubsub/GooglePubSubMessageBinding.scala [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import amf.core.internal.metamodel.Field
+import amf.core.internal.parser.domain.{Annotations, Fields}
+import amf.shapes.client.scala.model.domain.Key
+
+class GooglePubSubMessageBinding(override val fields: Fields, override val annotations: Annotations)
+    extends MessageBinding
+    with BindingVersion
+    with Key {
+
+  override protected def bindingVersionField: Field               = BindingVersion
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/client/scala/model/domain/bindings/googlepubsub/GooglePubSubMessageBinding.scala:22]: Should return StrField, not Field

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/metamodel/domain/bindings/GooglePubSubBindingModel.scala [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+    )
+
+  val OrderingKey: Field =
+    Field(
+      Str,
+      ApiBinding + "orderingKey",
+      ModelDoc(
+        ModelVocabularies.ApiBinding,
+        "orderingKey",
+        "identifies related messages for which publish order should be respected"
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/internal/metamodel/domain/bindings/GooglePubSubBindingModel.scala]: Typo in description

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/metamodel/domain/bindings/GooglePubSubBindingModel.scala [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+    )
+
+  val SchemaSettings: Field =
+    Field(
+      GooglePubSubSchemaSettingsModel,
+      ApiBinding + "schemaSettings",
+      ModelDoc(
+        ModelVocabularies.ApiBinding,
+        "messageStoragePolicy",
+        "Policy constraining the set of Google Cloud Platform regions where messages published to the topic may be stored"
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/internal/metamodel/domain/bindings/GooglePubSubBindingModel.scala]: Wrong description

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/metamodel/domain/bindings/GooglePubSubBindingModel.scala [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+    )
+
+  val Name: Field =
+    Field(
+      Str,
+      ApiBinding + "name",
+      ModelDoc(
+        ModelVocabularies.ApiBinding,
+        "name",
+        "TThe name of the schema that messages published should be validated against"
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/internal/metamodel/domain/bindings/GooglePubSubBindingModel.scala]: Typo in description

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/emitters/bindings/AsyncApiChannelBindingsEmitter.scala [hunk 10]
```diff
@@ -262,6 +281,84 @@ class AnypointMQChannelBindingEmitter(binding: AnypointMQChannelBinding, orderin
...
+      YNode(GooglePubSub),
+      _.obj { emitter =>
+        val result = ListBuffer[EntryEmitter]()
+        val fs     = binding.fields
+
+        fs.entry(GooglePubSubChannelBindingModel.Labels)
+          .foreach(f =>
+            result += EntryPartEmitter(
+              "labels",
+              DataNodeEmitter(binding.labels.asInstanceOf[ObjectNode], ordering)(spec.eh),
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/emitters/bindings/AsyncApiChannelBindingsEmitter.scala]: Unnecessary casting

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/emitters/bindings/AsyncApiMessageBindingsEmitter.scala [hunk 11]
```diff
@@ -179,3 +196,51 @@ class AnypointMQMessageEmitter(binding: AnypointMQMessageBinding, ordering: Spec
...
+      YNode(GooglePubSub),
+      _.obj { emitter =>
+        val result = ListBuffer[EntryEmitter]()
+        val fs     = binding.fields
+
+        fs.entry(GooglePubSubMessageBindingModel.Attributes)
+          .foreach(f =>
+            result += EntryPartEmitter(
+              "attributes",
+              DataNodeEmitter(binding.attributes.asInstanceOf[ObjectNode], ordering)(spec.eh),
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/emitters/bindings/AsyncApiMessageBindingsEmitter.scala]: Unnecessary casting

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/parser/bindings/message/GooglePubSubMessageBindingParser.scala [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    map.key(
+      "schema",
+      { entry =>
+        val schema    = GooglePubSubSchemaDefinition(Annotations(entry.value))
+        val schemaMap = entry.value.as[YMap]
+
+        schemaMap.key("name", GooglePubSubSchemaDefinitionModel.Name in schema)
+        schemaMap.key("fieldType", GooglePubSubSchemaDefinitionModel.FieldType in schema)
+
+        ctx.closedShape(schema, schemaMap, "GooglePubSubMessageSchema")
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/parser/bindings/message/GooglePubSubMessageBindingParser.scala:51]: Missing close shape

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/parser/context/syntax/Async25Syntax.scala [hunk 13]
```diff
@@ -1,11 +1,36 @@
...
+      "encoding",
+      "firstRevisionId",
+      "lastRevisionId",
+      "name"
+    ),
+    "GooglePubSubMessageBinding" -> Set(
+      "attributes",
+      "orderingKey",
+      "schema",
+      "bindingVersion"
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/parser/context/syntax/Async25Syntax.scala:33]: Missing close shape GooglePubSubMessageSchema

File: amf-cli/shared/src/test/resources/upanddown/cycle/async20/bindings/gps-message-binding.yaml [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+            type: string
+            description: Unique identifier for the user.
+          action:
+            type: string
+            description: Description of the user action.
+      bindings:
+        googlepubsub:
+          attributes:
+            exampleAttribute: exampleValue
+          orderingKey: userEventOrder
```
Comment (looseale) [inline amf-cli/shared/src/test/resources/upanddown/cycle/async20/bindings/gps-message-binding.yaml:31]: Add `schema` key to the test. We should test all the fields

File: amf-cli/shared/src/test/resources/validations/async20/validations/binding-wrong-key.yaml [hunk 15]
```diff
@@ -0,0 +1,35 @@
+asyncapi: 2.5.0
+info:
+  title: API with GooglePubSub Message Binding Validation
+  version: 1.0.0
+servers:
+  googlePubSubServer:
+    url: localhost:9092
+    protocol: googlepubsub
+    description: A Google Pub/Sub server for messaging.
+    extraKey: an extra key
```
Comment (looseale) [inline amf-cli/shared/src/test/resources/validations/async20/validations/binding-wrong-key.yaml]: Not necessary to test this. Server node is not new.

File: amf-cli/shared/src/test/resources/validations/async20/validations/binding-wrong-key.yaml [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+            type: string
+            description: Unique identifier for the user.
+          action:
+            type: string
+            description: Description of the user action.
+      bindings:
+        googlepubsub:
+          attributes:
+            exampleAttribute: "exampleValue"
+          wrongOrderingKey: "userEventOrder"
```
Comment (looseale) [inline amf-cli/shared/src/test/resources/validations/async20/validations/binding-wrong-key.yaml]: Try also adding extra keys to new nested complex models, like `schema` in this case

File: amf-cli/shared/src/test/resources/upanddown/cycle/async20/bindings/gps-message-binding.yaml (hunk 14 above)
Comment (looseale) [reply amf-cli/shared/src/test/resources/upanddown/cycle/async20/bindings/gps-message-binding.yaml:31]: Also, add one ChannelBinding testing the same, in the 3 new models

**Review by looseale — CHANGES_REQUESTED** [review]

File: amf-api-contract/shared/src/main/scala/amf/apicontract/client/platform/model/domain/bindings/googlepubsub/GooglePubSubMessageBinding.scala (hunk 2 above)
Comment (damianpedra) [reply amf-api-contract/shared/src/main/scala/amf/apicontract/client/platform/model/domain/bindings/googlepubsub/GooglePubSubMessageBinding.scala]: fixed

File: amf-api-contract/shared/src/main/scala/amf/apicontract/client/platform/model/domain/bindings/googlepubsub/GooglePubSubMessageBinding.scala (hunk 3 above)
Comment (damianpedra) [reply amf-api-contract/shared/src/main/scala/amf/apicontract/client/platform/model/domain/bindings/googlepubsub/GooglePubSubMessageBinding.scala]: done

File: amf-api-contract/shared/src/main/scala/amf/apicontract/client/scala/model/domain/bindings/googlepubsub/GooglePubSubChannelBinding.scala (hunk 4 above)
Comment (damianpedra) [reply amf-api-contract/shared/src/main/scala/amf/apicontract/client/scala/model/domain/bindings/googlepubsub/GooglePubSubChannelBinding.scala]: done

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/parser/bindings/message/GooglePubSubMessageBindingParser.scala (hunk 12 above)
Comment (damianpedra) [reply amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/parser/bindings/message/GooglePubSubMessageBindingParser.scala:51]: close shape added and tested

File: amf-api-contract/shared/src/main/scala/amf/apicontract/client/scala/model/domain/bindings/googlepubsub/GooglePubSubChannelBinding.scala (hunk 5 above)
Comment (damianpedra) [reply amf-api-contract/shared/src/main/scala/amf/apicontract/client/scala/model/domain/bindings/googlepubsub/GooglePubSubChannelBinding.scala:17]: As we saw binding version should be Field

File: amf-cli/shared/src/test/resources/upanddown/cycle/async20/bindings/gps-message-binding.yaml (hunk 14 above)
Comment (damianpedra) [reply amf-cli/shared/src/test/resources/upanddown/cycle/async20/bindings/gps-message-binding.yaml:31]: added all schemas to test in message and channels

File: amf-cli/shared/src/test/resources/validations/async20/validations/binding-wrong-key.yaml (hunk 15 above)
Comment (damianpedra) [reply amf-cli/shared/src/test/resources/validations/async20/validations/binding-wrong-key.yaml]: deleted

File: amf-cli/shared/src/test/resources/validations/async20/validations/binding-wrong-key.yaml (hunk 16 above)
Comment (damianpedra) [reply amf-cli/shared/src/test/resources/validations/async20/validations/binding-wrong-key.yaml]: added schema

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/parser/context/syntax/Async25Syntax.scala (hunk 13 above)
Comment (damianpedra) [reply amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/parser/context/syntax/Async25Syntax.scala:33]: close shape already added

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/emitters/bindings/AsyncApiChannelBindingsEmitter.scala (hunk 10 above)
Comment (damianpedra) [reply amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/emitters/bindings/AsyncApiChannelBindingsEmitter.scala]: deleted

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/emitters/bindings/AsyncApiMessageBindingsEmitter.scala (hunk 11 above)
Comment (damianpedra) [reply amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/emitters/bindings/AsyncApiMessageBindingsEmitter.scala]: deleted

**Review by looseale — APPROVED** [review]

**Review by nschejtman — APPROVED** [review]

**Pull request merged**